### PR TITLE
issue #180 - changing download approach and adding progress indicator

### PR DESCRIPTION
**Pull request by bfitzpat**
Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

### CODE REVIEW

**Comment by bfitzpat** [comment]
This puts the download progress in the Apache Camel K output channel.

```
Installing kamel... 
Downloading kamel cli tool from https://github.com/apache/camel-k/releases/download/1.0.0-M3/camel-k-client-1.0.0-M3-linux-64bit.tar.gz to /home/bfitzpat/.config/Code/User/globalStorage/redhat.vscode-camelk/camelk/tools/kamel/kamel 
Downloading from: https://github.com/apache/camel-k/releases/download/1.0.0-M3/camel-k-client-1.0.0-M3-linux-64bit.tar.gz
Bytes to transfer: 23212001
Bytes transferred: 0
Bytes transferred: 16360
Bytes transferred: 16927
Bytes transferred: 33287
Bytes transferred: 34335
Bytes transferred: 50695
...
Bytes transferred: 23165087
Bytes transferred: 23165471
Bytes transferred: 23181831
Bytes transferred: 23198191
Bytes transferred: 23212001
Downloaded camel-k-client-1.0.0-M3-linux-64bit.tar.gz.
Decompressing camel-k-client-1.0.0-M3-linux-64bit.tar.gz.
done 
```

and

```
Installing kubectl... 
Downloading Kubernetes cli tool from https://storage.googleapis.com/kubernetes-release/release/v1.16.2/bin/linux/amd64/kubectl to /home/bfitzpat/.vs-kubernetes/tools/kubectl/kubectl 
Downloading from: https://storage.googleapis.com/kubernetes-release/release/v1.16.2/bin/linux/amd64/kubectl
Bytes to transfer: 46681472
Bytes transferred: 0
Bytes transferred: 466
Bytes transferred: 1855
Bytes transferred: 3244
Bytes transferred: 4633
...
Bytes transferred: 46644328
Bytes transferred: 46660712
Bytes transferred: 46677096
Bytes transferred: 46681472
Downloaded kubectl.
Decompressing kubectl.
done 
```

**Comment by bfitzpat** [comment]
Investigating moving to a progress bar as suggested by @lhein  -- https://github.com/jboss-fuse/vscode-atlasmap/blob/master/src/extension.ts#L215-L237

**Comment by bfitzpat** [comment]
Moved it to the status bar with a percentage rather than putting progress into the Output channel. Tried using the progress bar but had trouble getting that approach to work for me.

**Comment by lhein** [comment]
Why that? It worked fine yesterday except that the percentages where not displayed correctly due to some logical error. What made you switch to a static progress message instead?

File: src/installer.ts [hunk 1]
```diff
@@ -27,12 +27,11 @@ import * as shell from './shell';
...
 import * as kubectlutils from './kubectlutils';
 import * as downloader from './downloader';
-
-const downloadTarball = require('download-tarball');
+import * as download from 'download';
 
 export const kamel = 'kamel';
 export const kamel_windows = 'kamel.exe';
-export const version = '1.0.0-M2'; //need to retrieve this if possible, but have a default
+export const version = '1.0.0-M3'; //need to retrieve this if possible, but have a default
```
Comment (apupier) [inline src/installer.ts]: would be better to have a dedicated PR for default version upgrade.

File: src/installer.ts [hunk 2]
```diff
@@ -99,6 +98,43 @@ export function getPlatform() : string | undefined {
     return undefined;
 }
 
+async function downloadAndExtract(link : string, dlFilename: string, installFolder : string, extractFlag : boolean) : Promise<boolean> {
```
Comment (apupier) [inline src/installer.ts]: tabs and space mixed

File: src/installer.ts [hunk 3]
```diff
@@ -126,19 +162,20 @@ export async function installKamel(context: vscode.ExtensionContext): Promise<Er
...
+	await downloadAndExtract(kamelUrl, kamelCliFile, installFolder, true)
 	.then( async (flag) => {
-		console.log(`Downloaded ${downloadFile} successfully: ${flag}`);
-		if (fs.existsSync(downloadFile)) {
-			if (shell.isUnix()) {
-				fs.chmodSync(downloadFile, '0700');
-			}
-			await config.addKamelPathToConfig(downloadFile);
-		}
+	 	console.log(`Downloaded ${downloadFile} successfully: ${flag}`);
```
Comment (apupier) [inline src/installer.ts]: space added?

File: src/installer.ts [hunk 4]
```diff
@@ -233,14 +252,17 @@ export async function installKubectl(context: vscode.ExtensionContext): Promise<
...
+		console.log(`Downloaded ${downloadFile} successfully: ${flag}`);
+	}).catch ( (error) => {
+		console.log(error);
+		return { succeeded: false, error: [`Failed to download kubectl: ${error}`] };
+	});
 	await config.addKubectlPathToConfig(downloadFile);
 
 	if (shell.isUnix()) {
-		fs.chmodSync(downloadFile, '0777');
+		fs.chmodSync(downloadFile, '0700');
```
Comment (apupier) [inline src/installer.ts]: why do you need to change rights when changing the download approach?
If giving a 700, won't it means that the user needs to be an admin to be able to launch it? or that he/she launched VS Code in admin mode?

Comment (lhein) [reply src/installer.ts]: Nope....it means OWNER/USER is allowed everything, GROUP is allowed nothing, OTHERS is allowed nothing. So that change basically improves the security because it doesn't allow other users or groups to access the users home directory where this file is stored.

Comment (apupier) [reply src/installer.ts]: ah ok, seems better effectively.
remains my question:

why do you need to change rights when changing the download approach?

Comment (lhein) [reply src/installer.ts]: Why can't u ask the question you actually want to ask? "Can you make this change in a separate PR?" ;)

Comment (apupier) [reply src/installer.ts]: because if the answer is that we need this right modification to have the download progress working, it should not be in another PR. I don't think so but I'm not sure so I'm asking.

**Comment by lhein** [comment]
Brian, I see another issue. The download works fine but it repeats with every startup of vscode. It seems you only check if kubectl / kamel is available on PATH variable but you do not check if those executables have been downloaded before and are residing inside that users home folder somewhere. I think that needs to be addressed. (maybe in a separate issue/pr - just let me know what you prefer)

**Comment by lhein** [comment]
Brian, I created a PR on your fork which refactors this PR and switches back to using Progress Bar with correct percentage calculation and progress display. 
see https://github.com/bfitzpat/vscode-camelk/pull/4
Please check it and see if that works for you. Here its working perfectly.

**Comment by bfitzpat** [comment]
> Please check it and see if that works for you. Here its working perfectly.

The reason I moved away from this code was because of the ENOENT issue, which I was unable to work around. I honestly don't see the progress bar being an improvement, but beyond that it simply doesn't work on my box. See this - https://github.com/bfitzpat/vscode-camelk/pull/4#issuecomment-550321134

**Comment by bfitzpat** [comment]
> Brian, I see another issue. The download works fine but it repeats with every startup of vscode. It seems you only check if kubectl / kamel is available on PATH variable but you do not check if those executables have been downloaded before and are residing inside that users home folder somewhere. I think that needs to be addressed. (maybe in a separate issue/pr - just let me know what you prefer)

That is a separate issue, but I have seen it. It was a problem before this progress change.

File: src/installer.ts (hunk 4 above)
Comment (bfitzpat) [reply src/installer.ts]: @lhein helped change from 0777 to 0700 a while back for the kamel CLI and I just never changed it for the kubernetes side. Since we are working in this code, it did not seem a stretch to change it at this time. I honestly don't see the need to break this out

Comment (apupier) [reply src/installer.ts]: advantages:
- smaller PR
- this modification will no tbe blocked by user comments on the progress change
- allows to have a better tracing of modifications in history
- easier to think to add an entry in changelog
- easier to understand that the modification of the permission is not needed for the progress
drawbacks:
- extra branch and PR to create

Comment (bfitzpat) [reply src/installer.ts]: @apupier it's one line in the context of another change that still affects the same code. it's staying put.

**Comment by bfitzpat** [comment]
Shifted back to the progress bar after @lhein finally got it to work. Thanks Lars!

File: src/installer.ts [hunk 5]
```diff
@@ -93,10 +92,48 @@ export function getPlatform() : string | undefined {
...
+						lastProg = progressamt.transferred;
+					}
+					if (progressamt.transferred >= progressamt.total) {
+						success = true;
+					}
+				});
+			resolve(success);
+		})
+		.then( () => {
+			extension.mainOutputChannel.appendLine('done!');
```
Comment (apupier) [inline src/installer.ts]: as the channel is used to log different things, i think we should specify what is done.

File: src/installer.ts [hunk 6]
```diff
@@ -93,10 +92,48 @@ export function getPlatform() : string | undefined {
...
+}
+
+function downloadAndExtract(link : string, dlFilename: string, installFolder : string, extract : boolean) : Promise<boolean> {
+	return new Promise<boolean>( async (resolve, reject) => {
+		const downloadSettings = {
+			filename: `${dlFilename}`,
+			extract: extract,
+		};
+		extension.mainOutputChannel.appendLine('Downloading from: ' + link);
+		let progressTitle : string = `Download (${dlFilename})`;
```
Comment (apupier) [inline src/installer.ts]: What do you think to use "Downloading" instead of "Download"?

**Comment by bfitzpat** [comment]
Though it worked locally, it does not work on travis - I reverted the change and went back to the status bar.

File: src/installer.ts [hunk 7]
```diff
@@ -93,10 +92,47 @@ export function getPlatform() : string | undefined {
...
+		extract: extractFlag,
+	  };
+	extension.mainOutputChannel.appendLine('Downloading from: ' + link);
+	await download(link, installFolder, downloadSettings)
+		.on('response', (response) => {
+			extension.mainOutputChannel.appendLine(`Bytes to transfer: ${response.headers['content-length']}`);
+		}).on('downloadProgress', (progress) => {
+			let incr = progress.total > 0 ? Math.floor(progress.transferred / progress.total * 100) : 0;
+			let percent = Math.round(incr);
+			let message = `Download progress: ${progress.transferred} / ${progress.total} (${percent}%)`;
```
Comment (apupier) [inline src/installer.ts]: I know the available space is reduced in the bottom bar but indicating what is downloaded would be nice. Maybe adding it as a tooltip is a good alternative?

**Comment by apupier** [comment]
> > Brian, I see another issue. The download works fine but it repeats with every startup of vscode. It seems you only check if kubectl / kamel is available on PATH variable but you do not check if those executables have been downloaded before and are residing inside that users home folder somewhere. I think that needs to be addressed. (maybe in a separate issue/pr - just let me know what you prefer)
> 
> That is a separate issue, but I have seen it. It was a problem before this progress change.

reported https://issues.jboss.org/browse/FUSETOOLS2-52 to handle this case

File: src/installer.ts (hunk 7 above)
Comment (bfitzpat) [reply src/installer.ts]: Added tooltip

**Comment by bfitzpat** [comment]
@lhein I'm going to go ahead and merge this, but if you want to provide a PR later that shifts us back to a progress bar, that would be fantastic.
